modules/listener/listener.bal

```diff
@@ -0,0 +1,118 @@
+
+    public isolated function init(int port, ListenerConfiguration config) {
+        self.httpListener = new (port);
+        self.token = config.verificationToken;
+    }
+
+    public isolated function attach(http:Service s, string[]|string? name) returns error? {
+        return self.httpListener.attach(s, name);
+    }
+
+    public isolated function detach(http:Service s) returns error? {
+        return self.httpListener.detach(s);
+    }
+
+    public isolated function 'start() returns error? {
+        return self.httpListener.'start();
+    }
+
+    public isolated function gracefulStop() returns error? {
+        return ();
+    }
+
+    public isolated function immediateStop() returns error? {
+        return self.httpListener.immediateStop();
+    }
+
+    # The `listener.getEventData(http:Caller caller, http:Request request)` function processes 
+    #
+    # + caller - http:Caller for acknowleding to the events received
+    # + slackRequest - http:Request which contains all the event related data
+    # + return - A `error` if it is a failure or the `SlackEvent` record if it is a success
+    public isolated function getEventData(http:Caller caller, http:Request slackRequest) returns @untainted error|SlackEvent {
+        json request = check slackRequest.getJsonPayload();
+        //calculation to identify the delay of event receiving from Slack
+        int slackTimeStamp = check 'int:fromString(slackRequest.getHeader(HEADER_TIMESTAMP));
+        int nowTimeStamp = check 'int:fromString(time:currentTime().time.toString().substring(0, 10));
+        int timeDiff = nowTimeStamp - slackTimeStamp;
+        //validate the request by comparing the token received with your app token and by checking whether the timestamp is withing 5 minutes (60 * 5 seconds) range
+        if (request.token == self.token && timeDiff < 60 * 5) {
+            string reqType = request.'type.toString();
+            if (reqType == URL_VERIFICATION) {
+                return self.verifyURL(caller, request);
+            } else if (reqType == EVENT_CALLBACK) {
+                return self.getEventCallBackData(caller, request);
+            } else {
+                return error("Unidentified Request Type");
+            }
+        } else {
+            check caller->respond(http:STATUS_BAD_REQUEST);
+            return error("Invalid Request");
+        }
+    }
+
+    isolated function verifyURL(http:Caller caller, json req) returns @untainted error|SlackEvent {
+        http:Response response = new;
+        ValidationRequest validationRqst = check req.cloneWithType(ValidationRequest);
+        response.statusCode = http:STATUS_OK;
+        response.setPayload({challenge: <@untainted>validationRqst.challenge});
+        check caller->respond(response);
+        log:print("Request URL Verified");
+        return validationRqst;
+    }
+
+    isolated function getEventCallBackData(http:Caller caller, json req) returns @untainted error|SlackEvent {
```

Comment:
Please review this function

Reply:
As you have used this ```public type SlackEvent GenericSlackEvent|ValidationRequest|InviteRequestedEvent|AppEvent|CallEvent|MessageEvent|FileEvent|DNDEvent|ReactionEvent|MemberEvent;```
We can use `typedesc` to check the record type;
Say : we have `SlackEvent| MessageEventType|FileEventType eventWegot;` and it is really `MessageEvent`;
We can use in side if  `eventWegot is typedesc<MessageEventType>` as you have used types.
 then clone to the MessageEvent;

Reply:
As per the discussion we had this approach does not match with our usecase.